consul/leader.go

```diff
@@ -562,11 +562,36 @@ func (s *Server) joinConsulServer(m serf.Member, parts *agent.Server) error {
 	}
 
 	// Attempt to add as a peer
-	addFuture := s.raft.AddPeer(raft.ServerAddress(addr))
-	if err := addFuture.Error(); err != nil {
-		s.logger.Printf("[ERR] consul: failed to add raft peer: %v", err)
+	minRaftProtocol, err := ServerMinRaftProtocol(s.serfLAN.Members())
+	if err != nil {
 		return err
 	}
+
+	if minRaftProtocol >= 2 && parts.RaftVersion >= 3 {
```

Comment:
I think we need to adjust the logic like this:

```
if parts.RaftVersion >= 3 {
    if minRaftProtocol >= 2 {
        AddVoter
    } else {
        // Log an error that we can't add a Raft version 3 peer since the cluster hasn't been upgraded yet.
} else {
    AddPeer
}
```

That way we protect the cluster if they try to add a server that's too new to interact with old servers in the cluster.